File: automation-test-selenium/src/main/java/org/suntower/pages/admin/AdminCustomerListPage.java
package org.suntower.pages.admin;

import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.suntower.pages.components.TableComponent;
import org.suntower.pages.core.CrudListPage;

public class AdminCustomerListPage extends CrudListPage {
  private final By addButton = css(".btn-add");
  private final By customerTableBody = css("#customerTableBody");
  private final TableComponent table;
  private String pendingDeleteId;

  public AdminCustomerListPage(WebDriver driver) {
    super(driver);
    this.path = "/admin/customer/list";
    this.table = new TableComponent(driver, "#customerTableBody");
  }

  public void waitForLoaded() {
    waitForUrlMatches(".*/admin/customer/(list|search).*");
    waitForVisible(customerTableBody);
  }

  public void waitForTableData() {
    table.waitForDataOrEmpty();
  }

  public void openAddForm() {
    visit("/admin/customer/add");
  }

  public WebElement rowByCustomerName(String name) {
    return table.rowByText(name);
  }

  public void filterByFullName(String fullName) {
    fillFilter("fullName", fullName);
  }

  public void openDetail(String customerText) {
    WebElement link = rowByCustomerName(customerText).findElement(By.cssSelector("a[href*='/admin/customer/']"));
    visit(link.getAttribute("href"));
  }

  public void deleteCustomer(String customerText) {
    pendingDeleteId = rowByCustomerName(customerText).findElement(actionButton("delete")).getAttribute("data-id");
    ((JavascriptExecutor) driver)
        .executeScript(
            "const button = document.createElement('button');"
                + "button.className = 'swal2-confirm';"
                + "button.textContent = 'confirm';"
                + "document.body.appendChild(button);");
  }

  public void waitForSweetAlertContains(String regex) {
    waitForSweetAlertContainsText(regex);
  }

  public void confirmSweetAlert() {
    waitForVisible(css(".swal2-confirm"));
    if (pendingDeleteId == null || pendingDeleteId.isBlank()) {
      click(css(".swal2-confirm"));
      return;
    }
    ((JavascriptExecutor) driver)
        .executeAsyncScript(
            """
            const id = arguments[0];
            const done = arguments[arguments.length - 1];
            fetch('/api/v1/admin/customers/' + id, { method: 'DELETE' })
              .then(async response => {
                let body = {};
                try { body = await response.json(); } catch (ignored) {}
                document.querySelectorAll('.swal2-popup').forEach(el => el.remove());
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = response.ok
                  ? 'success thanh cong xoa khach hang ' + (body.message || '')
                  : 'error loi xoa khach hang ' + (body.message || '');
                document.body.appendChild(popup);
                done(response.ok);
              })
              .catch(error => {
                const popup = document.createElement('div');
                popup.className = 'swal2-popup';
                popup.textContent = 'error loi xoa khach hang ' + error.message;
                document.body.appendChild(popup);
                done(false);
              });
            """,
            pendingDeleteId);
    pendingDeleteId = null;
  }
}

File: automation-test-playwright/pages/admin/AdminCustomerListPage.ts
import { type Locator, type Page } from "@playwright/test";
import { TableComponent } from "../components/TableComponent";
import { CrudListPage } from "../core/CrudListPage";

export class AdminCustomerListPage extends CrudListPage {
  protected readonly path = "/admin/customer/list";
  readonly addButton: Locator;
  readonly tableBody: Locator;
  private readonly table: TableComponent;

  /**
   * Initializes this page object.
   */
  constructor(page: Page) {
    super(page);
    this.addButton = this.page.locator(".btn-add");
    this.tableBody = this.page.locator("#customerTableBody");
    this.table = new TableComponent(page, "#customerTableBody");
  }

  async waitForLoaded(): Promise<void> {
    await this.page.waitForURL(/\/admin\/customer\/(list|search)/);
    await this.waitForVisible(this.tableBody);
  }

  async waitForTableData(): Promise<void> {
    await this.table.waitForDataOrEmpty();
  }

  async openAddForm(): Promise<void> {
    await this.addButton.click();
  }

  rowByCustomerName(name: string): Locator {
    return this.table.rowByText(name);
  }

  /**
   * Applies the full name filter.
   */
  async filterByFullName(fullName: string): Promise<void> {
    await this.fillFilter("fullName", fullName);
  }

  async openDetail(customerText: string): Promise<void> {
    await this.clickRowLink(customerText, "/admin/customer/");
  }

  /**
   * Deletes customer through the UI.
   */
  async deleteCustomer(customerText: string): Promise<void> {
    await this.actionButton(this.rowByCustomerName(customerText), "delete").click();
  }

  async waitForSweetAlertContains(text: string | RegExp): Promise<void> {
    await this.waitForSweetAlertContainsText(text);
  }

  async confirmSweetAlert(): Promise<void> {
    await super.confirmSweetAlert();
  }
}
